File: manager/activation-issuer/src/main/java/com/Night/night/issuer/MainActivity.java
package com.Night.night.issuer;

import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Base64;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.Signature;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.Locale;

public final class MainActivity extends Activity {
    private EditText androidId;
    private TextView activationCode;
    private TextView keyStatus;
    private byte[] privateKeyDer;

    @Override public void onCreate(Bundle state) {
        super.onCreate(state);
        float density = getResources().getDisplayMetrics().density;
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setGravity(Gravity.CENTER_HORIZONTAL);
        root.setPadding((int)(24*density), (int)(48*density), (int)(24*density), (int)(24*density));
        root.setBackgroundColor(Color.rgb(7, 11, 20));

        TextView title = text("NIGHT / ACTIVATION ISSUER", 24, Color.rgb(153, 211, 255));
        root.addView(title);
        root.addView(text("输入目标设备 Android ID，离线签发专属激活码", 14, Color.rgb(160, 174, 200)));
        androidId = new EditText(this);
        androidId.setHint("Android ID");
        androidId.setTextColor(Color.WHITE);
        androidId.setHintTextColor(Color.GRAY);
        root.addView(androidId, wide());
        keyStatus = text("正在载入内置 Night 私钥…", 13, Color.rgb(160, 174, 200));
        root.addView(keyStatus, wide());
        Button issue = new Button(this);
        issue.setText("签发激活码");
        issue.setOnClickListener(v -> issue());
        root.addView(issue, wide());
        activationCode = text("", 13, Color.rgb(216, 255, 225));
        activationCode.setTextIsSelectable(true);
        activationCode.setPadding(0, (int)(18*density), 0, (int)(18*density));
        root.addView(activationCode, wide());
        Button copy = new Button(this);
        copy.setText("复制激活码");
        copy.setOnClickListener(v -> {
            if (activationCode.getText().length() == 0) return;
            getSystemService(ClipboardManager.class).setPrimaryClip(
                ClipData.newPlainText("Night activation code", activationCode.getText()));
            Toast.makeText(this, "已复制", Toast.LENGTH_SHORT).show();
        });
        root.addView(copy, wide());
        setContentView(root);
        loadBundledPrivateKey();
    }

    private void issue() {
        try {
            String id = androidId.getText().toString().trim().toLowerCase(Locale.ROOT);
            if (id.isEmpty()) throw new IllegalArgumentException("Android ID 不能为空");
            if (privateKeyDer == null) throw new IllegalStateException("内置激活私钥不可用");
            PrivateKey key = KeyFactory.getInstance("RSA").generatePrivate(new PKCS8EncodedKeySpec(privateKeyDer));
            Signature signer = Signature.getInstance("SHA256withRSA");
            signer.initSign(key);
            signer.update(("Night|1|" + id).getBytes(StandardCharsets.UTF_8));
            activationCode.setText("N1." + Base64.encodeToString(signer.sign(), Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING));
        } catch (Throwable error) {
            activationCode.setText("签发失败：" + error.getMessage());
        }
    }

    private void loadBundledPrivateKey() {
        try {
            Throwable firstError = null;
            for (String asset : new String[] {
                "night_activation_private_key.pk8",
                "night_activation_private_key.pem"
            }) {
                try {
                    byte[] raw = readAsset(asset);
                    privateKeyDer = normalizePrivateKey(raw);
                    keyStatus.setText("内置 Night 私钥已载入（PEM / PK8）");
                    return;
                } catch (Throwable error) {
                    if (firstError == null) firstError = error;
                }
            }
            throw new IllegalStateException(firstError == null ? "密钥资源不存在" : firstError.getMessage());
        } catch (Throwable error) {
            privateKeyDer = null;
            keyStatus.setText("内置私钥载入失败：" + error.getMessage());
        }
    }

    private byte[] readAsset(String name) throws Exception {
        try (java.io.InputStream input = getAssets().open(name);
             java.io.ByteArrayOutputStream output = new java.io.ByteArrayOutputStream()) {
            byte[] buffer = new byte[8192];
            for (int read; (read = input.read(buffer)) >= 0;) {
                output.write(buffer, 0, read);
            }
            return output.toByteArray();
        }
    }

    private byte[] normalizePrivateKey(byte[] raw) throws Exception {
        String text = new String(raw, StandardCharsets.US_ASCII);
        if (text.contains("BEGIN PRIVATE KEY")) {
            text = text.replace("-----BEGIN PRIVATE KEY-----", "")
                .replace("-----END PRIVATE KEY-----", "").replaceAll("\\s", "");
            raw = Base64.decode(text, Base64.DEFAULT);
        }
        KeyFactory.getInstance("RSA").generatePrivate(new PKCS8EncodedKeySpec(raw));
        return raw;
    }

    private TextView text(String value, int sp, int color) {
        TextView view = new TextView(this);
        view.setText(value); view.setTextSize(sp); view.setTextColor(color); view.setPadding(0, 8, 0, 8);
        return view;
    }

    private ViewGroup.LayoutParams wide() {
        return new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
    }
}

File: manager/offline-patcher/src/main/java/com/Night/night/patcher/MainActivity.kt
package com.Night.night.patcher

import android.app.Activity
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.Bundle
import android.provider.DocumentsContract
import android.provider.OpenableColumns
import android.view.Gravity
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import java.io.File
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class MainActivity : Activity() {
    private val kmis = listOf(
        "android12-5.10", "android13-5.10", "android13-5.15", "android14-5.15",
        "android14-6.1", "android15-6.6", "android16-6.12", "android17-6.18",
    )
    private lateinit var sourceLabel: TextView
    private lateinit var outputLabel: TextView
    private lateinit var kmiSpinner: Spinner
    private lateinit var patchButton: Button
    private lateinit var restoreButton: Button
    private lateinit var logView: TextView
    private var sourceUri: Uri? = null
    private var outputTreeUri: Uri? = null
    private var sourceName = "boot.img"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.statusBarColor = Color.rgb(7, 11, 20)
        window.navigationBarColor = Color.rgb(7, 11, 20)
        val density = resources.displayMetrics.density
        fun dp(value: Int) = (value * density).toInt()

        val content = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(20), dp(28), dp(20), dp(28))
            setBackgroundColor(Color.rgb(7, 11, 20))
        }
        content.addView(TextView(this).apply {
            text = "NIGHT / MANUAL OUTPUT V1.3"
            textSize = 25f
            setTypeface(typeface, Typeface.BOLD)
            setTextColor(Color.rgb(153, 211, 255))
        })
        content.addView(TextView(this).apply {
            text = "LKM RAMDISK 修补 boot / init_boot · 手动选择输出目录"
            setTextColor(Color.rgb(152, 166, 194))
            setPadding(0, dp(8), 0, dp(20))
        })

        val card = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(18), dp(18), dp(18), dp(18))
            background = GradientDrawable().apply {
                shape = GradientDrawable.RECTANGLE
                cornerRadius = dp(24).toFloat()
                setColor(Color.argb(190, 24, 33, 51))
                setStroke(dp(1), Color.argb(130, 132, 196, 255))
            }
        }
        sourceLabel = label("尚未选择原始镜像", Color.WHITE)
        card.addView(sourceLabel)
        card.addView(actionButton("选择 boot.img / init_boot.img", Color.rgb(35, 93, 126)) { chooseSource() })
        outputLabel = label("输出目录：尚未选择", Color.rgb(141, 238, 213)).apply {
            setPadding(0, dp(10), 0, dp(12))
        }
        card.addView(outputLabel)
        card.addView(actionButton("手动选择输出目录", Color.rgb(38, 111, 91)) { chooseOutputDirectory() })
        card.addView(label("Night 专属 KMI", Color.WHITE))
        kmiSpinner = Spinner(this).apply {
            adapter = ArrayAdapter(this@MainActivity, android.R.layout.simple_spinner_dropdown_item, kmis)
            setSelection(4)
        }
        card.addView(kmiSpinner)
        patchButton = actionButton("开始脱机修补", Color.rgb(65, 105, 225)) { patch(false) }
        restoreButton = actionButton("移除 KernelSU / 恢复", Color.rgb(64, 75, 96)) { patch(true) }
        card.addView(patchButton)
        card.addView(restoreButton)
        content.addView(card)

        logView = TextView(this).apply {
            text = "V1.3 已就绪。请分别选择原始镜像和输出目录。"
            setTextColor(Color.rgb(199, 222, 247))
            setTextIsSelectable(true)
            setPadding(dp(14), dp(14), dp(14), dp(14))
            background = GradientDrawable().apply {
                cornerRadius = dp(18).toFloat()
                setColor(Color.rgb(14, 21, 34))
            }
        }
        content.addView(logView, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, dp(190)).apply {
            topMargin = dp(16)
        })
        content.addView(label("仅生成新镜像；不会覆盖原文件。", Color.rgb(132, 147, 173)).apply {
            gravity = Gravity.CENTER
            setPadding(0, dp(12), 0, 0)
        })
        setContentView(ScrollView(this).apply {
            addView(content, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        })
    }

    private fun label(value: String, color: Int) = TextView(this).apply {
        text = value
        setTextColor(color)
        textSize = 14f
    }

    private fun actionButton(value: String, color: Int, action: () -> Unit) = Button(this).apply {
        text = value
        backgroundTintList = ColorStateList.valueOf(color)
        setTextColor(Color.WHITE)
        setOnClickListener { action() }
    }

    private fun chooseSource() {
        startActivityForResult(Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = "application/octet-stream"
        }, REQUEST_SOURCE)
    }

    private fun chooseOutputDirectory() {
        startActivityForResult(Intent(Intent.ACTION_OPEN_DOCUMENT_TREE).apply {
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION or
                Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION or Intent.FLAG_GRANT_PREFIX_URI_PERMISSION)
        }, REQUEST_OUTPUT_DIRECTORY)
    }

    @Deprecated("Deprecated in Android")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != RESULT_OK || data?.data == null) return
        when (requestCode) {
            REQUEST_SOURCE -> {
                sourceUri = data.data
                runCatching {
                    contentResolver.takePersistableUriPermission(data.data!!, data.flags and
                        (Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION))
                }
                sourceName = queryDisplayName(data.data!!) ?: "boot.img"
                sourceLabel.text = "原始镜像：$sourceName"
                refreshOutputLabel()
            }
            REQUEST_OUTPUT_DIRECTORY -> {
                outputTreeUri = data.data
                runCatching {
                    contentResolver.takePersistableUriPermission(data.data!!, data.flags and
                        (Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION))
                }
                refreshOutputLabel()
            }
        }
    }

    private fun refreshOutputLabel() {
        val tree = outputTreeUri
        outputLabel.text = if (tree == null) {
            "输出目录：尚未选择"
        } else {
            val location = runCatching { DocumentsContract.getTreeDocumentId(tree) }.getOrDefault(tree.toString())
            "输出目录：$location\n文件名：${outputName(false)}"
        }
    }

    private fun patch(restore: Boolean) {
        val inputUri = sourceUri ?: run {
            Toast.makeText(this, "请先选择原始镜像", Toast.LENGTH_SHORT).show(); return
        }
        val outputTree = outputTreeUri ?: run {
            Toast.makeText(this, "请手动选择输出目录", Toast.LENGTH_SHORT).show(); return
        }
        patchButton.isEnabled = false
        restoreButton.isEnabled = false
        logView.text = "正在准备脱机资源…"
        Thread {
            try {
                val inputFile = File(cacheDir, "source.img")
                val moduleFile = File(cacheDir, "kernelsu.ko")
                val outputFile = File(cacheDir, if (restore) "Night_restored.img" else "Night_patched.img")
                contentResolver.openInputStream(inputUri)!!.use { input -> inputFile.outputStream().use(input::copyTo) }
                if (!restore) assets.open("kmi/${kmis[kmiSpinner.selectedItemPosition]}_kernelsu.ko").use { input ->
                    moduleFile.outputStream().use(input::copyTo)
                }
                outputFile.delete()
                val engine = File(applicationInfo.nativeLibraryDir, "libksud.so")
                val selectedKmi = kmis[kmiSpinner.selectedItemPosition]
                val args = if (restore) {
                    mutableListOf(engine.absolutePath, "boot-restore", "--boot", inputFile.absolutePath,
                        "--out", cacheDir.absolutePath, "--out-name", outputFile.name)
                } else {
                    // init_boot intentionally has no kernel. Use KernelSU's LKM ramdisk
                    // patcher for both boot and init_boot instead of the kernel-injection
                    // boot-patch-v2 command, which can only operate on boot images that
                    // contain a kernel block.
                    mutableListOf(engine.absolutePath, "boot-patch", "--boot", inputFile.absolutePath,
                        "--module", moduleFile.absolutePath, "--kmi", selectedKmi,
                        "--out", cacheDir.absolutePath, "--out-name", outputFile.name)
                }
                val process = ProcessBuilder(args).redirectErrorStream(true).start()
                val output = process.inputStream.bufferedReader().use { it.readText() }
                val exitCode = process.waitFor()
                check(exitCode == 0 && outputFile.isFile) { "修补失败，退出代码 $exitCode\n$output" }
                val outputUri = createManualOutput(outputTree, outputName(restore))
                contentResolver.openOutputStream(outputUri, "w")!!.use { target -> outputFile.inputStream().use { it.copyTo(target) } }
                val digest = sha256(outputFile)
                runOnUiThread {
                    outputLabel.text = "已输出：${outputName(restore)}\n位置：$outputUri"
                    logView.text = "模式：LKM RAMDISK（支持 init_boot）\n$output\n完成：$outputUri\nSHA-256：$digest"
                    Toast.makeText(this, "脱机处理完成", Toast.LENGTH_LONG).show()
                }
            } catch (error: Throwable) {
                runOnUiThread { logView.text = "错误：${error.message}"; Toast.makeText(this, "处理失败", Toast.LENGTH_LONG).show() }
            } finally {
                runOnUiThread { patchButton.isEnabled = true; restoreButton.isEnabled = true }
            }
        }.start()
    }

    private fun outputName(restore: Boolean): String {
        val base = sourceName.substringBeforeLast('.', sourceName).ifBlank { "boot" }
        return "${base}_Night_${if (restore) "restored" else "patched"}.img"
    }

    private fun queryDisplayName(uri: Uri): String? = contentResolver.query(
        uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null,
    )?.use { if (it.moveToFirst()) it.getString(0) else null }

    private fun createManualOutput(tree: Uri, name: String): Uri {
        check(DocumentsContract.isTreeUri(tree)) { "请选择有效的输出目录。" }
        val treeId = DocumentsContract.getTreeDocumentId(tree)
        val parent = DocumentsContract.buildDocumentUriUsingTree(tree, treeId)
        return checkNotNull(DocumentsContract.createDocument(contentResolver, parent, "application/octet-stream", name)) {
            "无法在所选目录创建输出文件。"
        }
    }

    private fun sha256(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buffer = ByteArray(1024 * 1024)
            while (true) { val read = input.read(buffer); if (read < 0) break; digest.update(buffer, 0, read) }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    companion object {
        private const val REQUEST_SOURCE = 1001
        private const val REQUEST_OUTPUT_DIRECTORY = 1002
    }
}
